terraform/bigquery.tf

```diff
@@ -362,6 +461,63 @@ resource "google_bigquery_table" "unique_events_view" {
   ]
 }
 
+# Unique Events -deduplicate rows but as a table function
+resource "google_bigquery_routine" "unique_events_by_date_type" {
+  dataset_id      = google_bigquery_dataset.default.dataset_id
+  routine_id      = "unique_events_by_date_type"
+  routine_type    = "TABLE_VALUED_FUNCTION"
+  language        = "SQL"
+  definition_body = <<EOT
+    SELECT
+      delivery_id,
+      signature,
+      received,
+      event,
+      payload,
+      JSON_VALUE(payload, "$.organization.login") organization,
+      SAFE_CAST(JSON_VALUE(payload, "$.organization.id") AS INT64) organization_id,
+      JSON_VALUE(payload, "$.repository.full_name") repository_full_name,
+      SAFE_CAST(JSON_QUERY(payload, "$.repository.id") AS INT64) repository_id,
+      JSON_VALUE(payload, "$.repository.name") repository,
+      JSON_VALUE(payload, "$.repository.visibility") repository_visibility,
+      JSON_VALUE(payload, "$.sender.login") sender,
+      SAFE_CAST(JSON_QUERY(payload, "$.sender.id") AS INT64) sender_id,
+    FROM
+      `${google_bigquery_dataset.default.dataset_id}.${google_bigquery_table.raw_events_table.table_id}`
+    WHERE
+      received >= start
+      AND received <= end
+      AND event = eventTypeFilter
+    GROUP BY
```

Comment:
Does it work to use SELECT DISTINCT and remove this huge GROUP BY clause?

https://cloud.google.com/bigquery/docs/reference/standard-sql/query-syntax#select_distinct

Reply:
Hmm, that may be possible, I'll look into it.

Reply:
I'll have to experiment with that

Reply:
Though as @lock14 pointed out in a chat thread, there are duplicates which we are not catching due to installation id being included in retried events that actually were successfully written, as payloads vary slightly.

I think that duplicate events may be better handled with a scheduled query than an TVF.

Reply:
I did update the existing view and this TVF to use SELECT_DISTINCT as in testing it was identical, thanks for the suggestion!